pkg/ipmanager/ipam/controller.go

```diff
@@ -0,0 +1,236 @@
+package ipam
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	k8sv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+)
+
+func NewController(spec SubmarinerIpamControllerSpecification, config InformerConfigStruct) *Controller {
+	exclusionMap := make(map[string]bool)
+	for _, v := range spec.ExcludeNS {
+		exclusionMap[v] = true
+	}
+
+	ipamController := &Controller{
+		kubeClientSet:    config.KubeClientSet,
+		serviceWorkqueue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "Services"),
+		servicesSynced:   config.ServiceInformer.Informer().HasSynced,
+		podWorkqueue:     workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "Pods"),
+		podsSynced:       config.PodInformer.Informer().HasSynced,
+
+		excludeNamespaces: exclusionMap,
+	}
+	klog.Info("Setting up event handlers")
+	config.ServiceInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc: ipamController.enqueueService,
+		UpdateFunc: func(old, new interface{}) {
+			ipamController.enqueueService(new)
+		},
+		DeleteFunc: ipamController.handleRemovedService,
+	}, 60*time.Second)
+	config.PodInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc: ipamController.enqueuePod,
+		UpdateFunc: func(old, new interface{}) {
+			ipamController.enqueuePod(new)
+		},
+		DeleteFunc: ipamController.handleRemovedPod,
+	}, 60*time.Second)
+
+	return ipamController
+}
+
+func (i *Controller) Run(stopCh <-chan struct{}) error {
+	var wg sync.WaitGroup
+	wg.Add(1)
+	defer utilruntime.HandleCrash()
+
+	// Start the informer factories to begin populating the informer caches
+	klog.Info("Starting IPAM Controller")
+
+	// Wait for the caches to be synced before starting workers
+	klog.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, i.servicesSynced, i.podsSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	klog.Info("Starting workers")
+	go wait.Until(i.runServiceWorker, time.Second, stopCh)
+	go wait.Until(i.runPodWorker, time.Second, stopCh)
+	wg.Wait()
+	<-stopCh
+	klog.Info("Shutting down workers")
+
+	return nil
+}
+
+func (i *Controller) runServiceWorker() {
+	for i.processNextService() {
+	}
+}
+
+func (i *Controller) runPodWorker() {
+	for i.processNextPod() {
+	}
+}
+
+const submarinerIpamGlobalIp = "submariner.io/globalIp"
```

Comment:
better at the top of the file or a constants file since it will probably be used for other things other than this controller

Reply:
coastguard follows this convention, but I'd prefer what @mpeterson suggested.